### PR TITLE
maint: Use `Storage` class hierarchy instead of `VariantStorage`

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference Issues/PRs

Towards #633.

To be merged after #670.

#### What does this implement/fix? How does it work (high level)? Highlight notable design decisions.

This PR:
 - removes `arcticdb::storage::VariantStorage` and perform associated name changes (note that the `arcticdb::proto::storage::VariantStorage` is left unchanged)
 - removes the CRTP on `arcticdb::storage::Storage`
 - defines pure virtual protected template methods
 - uses implementations of concrete classes extending `arcticdb::storage::Storage` as overrides of those template methods
 - moves `arcticdb::storage::s3::{S3,NfsBacked}Storage`'s implementation details in a dedicated file
 - simplifies the logic of `storage_specific` down to what we need (this needs to be discussed IMO)

#### Any other comments?

We could:
 - add inline documentation, although I do not think this is needed

We could in another PR:
 - rename `arcticdb::proto::storage::VariantStorage` to `arcticdb::proto::storage::StorageConfig` #709

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings and documentation?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

**Review by willdealtry — APPROVED** [review]
This is great, have wanted to see this done for a long time.

File: cpp/arcticdb/storage/storage.hpp [hunk 1]
```diff
@@ -80,34 +80,39 @@ class KeyNotFoundException : public ArcticSpecificException<ErrorCode::E_DUPLICA
...
 
     Storage(LibraryPath library_path, OpenMode mode) :
         lib_path_(std::move(library_path)),
         mode_(mode) {}
 
+    virtual ~Storage() = default;
+    Storage(const Storage&) = delete;
+    Storage& operator=(const Storage&) = delete;
+    Storage(Storage&&) = default;
+    Storage& operator=(Storage&&) = delete;
```
Comment (jjerphan) [inline cpp/arcticdb/storage/storage.hpp:94]: Note to reviewers: this does not entirely respect entity semantic (the move constructor is not deleted), but this is coherent with some other existing class declarations and definitions.

File: cpp/arcticdb/storage/s3/detail-inl.hpp
Comment (jjerphan) [inline cpp/arcticdb/storage/s3/detail-inl.hpp:1]: Note for reviewers: the content of this file originates from `cpp/arcticdb/storage/s3/s3_storage-inl.hpp`. Apart from potential simplification for headers' inclusion, it has not been modified.

File: cpp/arcticdb/storage/storage.hpp [hunk 2]
```diff
@@ -127,39 +132,58 @@ class Storage : public StorageBase {
...
+
+    virtual bool do_key_exists(const VariantKey& key) = 0;
+
+    virtual bool do_supports_prefix_matching() = 0;
+
+    virtual bool do_fast_delete() = 0;
+
+    virtual void do_iterate_type(KeyType key_type, const IterateTypeVisitor& visitor, const std::string & prefix) = 0;
+
+    virtual std::string do_storage_specific(const VariantKey& key) = 0;
```
Comment (jjerphan) [inline cpp/arcticdb/storage/storage.hpp:182]: Note for reviewers: as part of this PR, should we `const`-qualified methods which can? If so, should we also `const`-qualify their associated public method?

**Review by jjerphan — COMMENTED** [review]
I left some notes for reviewers.

File: cpp/arcticdb/storage/storage.hpp [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
 namespace arcticdb::storage {
 
-struct StorageBase {}; // marker class for type checking
+using ReadVisitor = std::function<void(const VariantKey&, Segment &&)>;
```
Comment (jjerphan) [inline cpp/arcticdb/storage/storage.hpp:28]: This was included in the last merge and this must be removed.

File: cpp/arcticdb/storage/azure/azure_storage.hpp [hunk 4]
```diff
@@ -23,11 +23,7 @@
 
 namespace arcticdb::storage::azure {
 
-class AzureStorage final : public Storage<AzureStorage> {
-
-    using Parent = Storage<AzureStorage>;
-    friend Parent;
-
+class AzureStorage final : public Storage {
```
Comment (DerThorsten) [inline cpp/arcticdb/storage/azure/azure_storage.hpp:26]: is there a technical reason to make it `final`?

File: cpp/arcticdb/storage/azure/azure_storage.hpp [hunk 5]
```diff
@@ -40,26 +36,28 @@ class AzureStorage final : public Storage<AzureStorage> {
     std::string get_key_path(const VariantKey& key) const;
 
   protected:
```
Comment (DerThorsten) [inline cpp/arcticdb/storage/azure/azure_storage.hpp:38]: `final` classes using `protected` makes no sense

File: cpp/arcticdb/storage/lmdb/lmdb_storage.hpp [hunk 6]
```diff
@@ -28,37 +28,35 @@ namespace fs = std::filesystem;
 
 namespace arcticdb::storage::lmdb {
 
-class LmdbStorage final : public Storage<LmdbStorage> {
-
-    using Parent = Storage<LmdbStorage>;
-    friend Parent;
-
+class LmdbStorage final : public Storage {
```
Comment (DerThorsten) [inline cpp/arcticdb/storage/lmdb/lmdb_storage.hpp:31]: why `final`

File: cpp/arcticdb/storage/lmdb/lmdb_storage.hpp [hunk 7]
```diff
@@ -28,37 +28,35 @@ namespace fs = std::filesystem;
...
   public:
     using Config = arcticdb::proto::lmdb_storage::Config;
 
     LmdbStorage(const LibraryPath &lib, OpenMode mode, const Config &conf);
 
   protected:
```
Comment (DerThorsten) [inline cpp/arcticdb/storage/lmdb/lmdb_storage.hpp:37]: `final` classes using `protected` seems pointless

File: cpp/arcticdb/storage/azure/azure_storage.hpp (hunk 4 above)
Comment (jjerphan) [reply cpp/arcticdb/storage/azure/azure_storage.hpp:26]: This indicates that concrete class aren't meant to be sub-classed.

I preserved this original choice, and have defined relevant method in concrete classes similarly.

File: cpp/arcticdb/storage/storage.hpp [hunk 8]
```diff
@@ -127,39 +130,57 @@ class Storage : public StorageBase {
...
-        return derived().do_key_exists(key);
+        return do_key_exists(key);
     }
 
     void iterate_type(KeyType key_type, const IterateTypeVisitor& visitor, const std::string &prefix = std::string()) {
-        derived().do_iterate_type(key_type, visitor, prefix);
+        do_iterate_type(key_type, visitor, prefix);
+    }
+
+    std::string storage_specific(const VariantKey& key) {
```
Comment (alexowens90) [inline cpp/arcticdb/storage/storage.hpp:156]: This is now a bad name. How about `key_path`?

File: cpp/arcticdb/storage/storage.hpp (hunk 2 above)
Comment (alexowens90) [reply cpp/arcticdb/storage/storage.hpp:180]: Makes sense to me, is there a reason not to that you can think of?

File: cpp/arcticdb/storage/storages.hpp [hunk 9]
```diff
@@ -125,30 +124,30 @@ class Storages {
...
-    std::vector<std::unique_ptr<VariantStorage>> variant_storages_;
+    std::vector<std::unique_ptr<Storage>> storages_;
     OpenMode mode_;
 };
 
 inline std::shared_ptr<Storages> create_storages(const LibraryPath& library_path, OpenMode mode, const arcticdb::proto::storage::VariantStorage &storage_config) {
-    using VariantVec = std::vector<std::unique_ptr<VariantStorage>>;
-    VariantVec variants;
-    variants.push_back(create_storage(library_path, mode, storage_config));
+    using StorageVec = std::vector<std::unique_ptr<Storage>>;
```
Comment (alexowens90) [inline cpp/arcticdb/storage/storages.hpp:137]: This and the same line in the following function seem pointless, can just use the definition from line 31? (I know you didn't add this, but while we're here we should remove anyway)

File: cpp/arcticdb/storage/storage.hpp (hunk 2 above)
Comment (jjerphan) [reply cpp/arcticdb/storage/storage.hpp:180]: I do not think of a reason not to use `const`-qualification when we actually can.

Comment (alexowens90) [reply cpp/arcticdb/storage/storage.hpp]: Yeh makes sense, no need to overcomplicate it right now

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**